inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp

```diff
@@ -431,13 +432,93 @@ void splitRow(const uint8_t* in, std::array<uint8_t*, chs>& outs, int length) {
     }
 }
 
+namespace {
+    template<typename type>
+    struct cv_type_to_depth;
+
+    template<> struct cv_type_to_depth<std::uint8_t>    { enum { depth = CV_8U  }; };
+    template<> struct cv_type_to_depth<std::int8_t>     { enum { depth = CV_8S  }; };
+    template<> struct cv_type_to_depth<std::uint16_t>   { enum { depth = CV_16U }; };
+    template<> struct cv_type_to_depth<std::int16_t>    { enum { depth = CV_16S }; };
+    template<> struct cv_type_to_depth<std::int32_t>    { enum { depth = CV_32S }; };
+    template<> struct cv_type_to_depth<float>           { enum { depth = CV_32F }; };
+
+    template<typename ... types>
+    struct typelist {};
+
+    template<typename type_list>
+    struct head;
+
+    template<template<typename ...> class list, typename head_t, typename ... types>
+    struct head<list<head_t, types...>> { using type = head_t;};
+
+    template<typename typelist>
+    using head_t = typename head<typelist>::type;
+
+    template<typename type>
+    struct type_to_type {};
+}
+
+namespace {
+    template <typename typelist>
+    struct type_dispatch_impl;
+
+    template <template<typename ...> class typelist, typename... type>
+    struct type_dispatch_impl<typelist<type...>> {
+
+        template <typename result_t, typename default_t, typename type_id_t, typename type_to_id_t, typename type_to_value_t>
+        static result_t dispatch(type_id_t type_id, type_to_id_t&& type_to_id, type_to_value_t&& type_to_value, default_t default_value) {
+            result_t res = default_value;
+
+            std::initializer_list<int> ({(type_id == type_to_id(type_to_type<type>{}) ? (res = type_to_value(type_to_type<type>{})), 0 : 0)...});
+            return res;
+        }
+    };
+}
+
+template<typename typelist, typename default_t, typename type_id_t, typename type_to_id_t, typename type_to_value_t,
+         typename result_t = decltype(std::declval<type_to_value_t>()(type_to_type<head_t<typelist>> {}))>
+result_t type_dispatch(type_id_t type_id, type_to_id_t&& type_to_id, type_to_value_t&& type_to_value, default_t default_value = {}){
```

Comment:
need to add comment what this tool is intended for and insight key moment in the implementation

Reply:
Yes please

Reply:
let's add after